### ISSUE TITLE
dotnet test can't find included runtime

### ISSUE THREAD

**Issue by livarcocc** [microsoft/vstest#2228]
_From @patearl on Tuesday, October 22, 2019 1:37:20 PM_

## Steps to reproduce
Create an nunit test project that targets netcoreapp3.0. As of today, it uses the latest microsoft test sdk package and the latest nunit adapter package.

Download the 3.0 SDK ZIP file and unzip it.

Run dotnet test from the 3.0 SDK's folder against the solution.

As can be seen below, running dotnet --list-runtimes shows the 3.0 runtime included with the SDK. However, when dotnet test is running, it is unable to find the same runtime that it just found and only lists the pre-installed runtimes. We would prefer not to actually install the runtime and just have it run properly from the zip archive SDK. It should be able to use the runtime that it is clearly able to find with --list-runtimes instead of only being able to use the pre-installed runtimes that it later lists during test execution.

## Expected  behavior
Runs tests

## Actual behavior
[22:42:47][Exec] Test run for C:\w\99694c0b2afac045\Source\Pleasant.Tests\bin\Release\netcoreapp3.0\Pleasant.Tests.dll(.NETCoreApp,Version=v3.0)
[22:42:47][Exec] Microsoft (R) Test Execution Command Line Tool Version 16.3.0
[22:42:47][Exec] Copyright (c) Microsoft Corporation.  All rights reserved.
[22:42:47][Exec] 
[22:42:47][Exec] Starting test execution, please wait...
[22:42:48][Exec] 
[22:42:48][Exec] A total of 1 test files matched the specified pattern.
[22:42:48][Exec] EXEC It was not possible to find any compatible framework version
[22:42:48][Exec] The specified framework 'Microsoft.NETCore.App', version '3.0.0' was not found.
[22:42:48][Exec]   - Check application dependencies and target a framework version installed at:
[22:42:48][Exec]       C:\Program Files\dotnet
[22:42:48][Exec]   - Installing .NET Core prerequisites might help resolve this problem:
[22:42:48][Exec]       https://go.microsoft.com/fwlink/?LinkID=798306&clcid=0x409
[22:42:48][Exec]   - The .NET Core framework and SDK can be installed from:
[22:42:48][Exec]       https://aka.ms/dotnet-download
[22:42:48][Exec]   - The following versions are installed:
[22:42:48][Exec]       1.0.1 at [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
[22:42:48][Exec]       2.0.5 at [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
[22:42:48][Exec]       2.0.7 at [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
[22:42:48][Exec]       2.0.9 at [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
[22:42:48][Exec]       2.1.3 at [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
[22:42:48][Exec]       2.1.5 at [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
[22:42:48][Exec]       2.1.9 at [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
[22:42:48][Exec]       2.1.11 at [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
[22:42:48][Exec]       2.1.12 at [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]

## Environment data
Using 3.0.100 SDK.

[22:42:25][TestSolution] Exec
[22:42:25][Exec] "C:\w\99694c0b2afac045\Build\dotnetsdk\dotnet" --list-sdks
[22:42:25][Exec] 1.0.0-preview2-003131 [C:\Program Files\dotnet\sdk]
[22:42:25][Exec] 2.1.4 [C:\Program Files\dotnet\sdk]
[22:42:25][Exec] 2.1.201 [C:\Program Files\dotnet\sdk]
[22:42:25][Exec] 2.1.202 [C:\Program Files\dotnet\sdk]
[22:42:25][Exec] 2.1.401 [C:\Program Files\dotnet\sdk]
[22:42:25][Exec] 2.1.403 [C:\Program Files\dotnet\sdk]
[22:42:25][Exec] 2.1.505 [C:\Program Files\dotnet\sdk]
[22:42:25][Exec] 2.1.507 [C:\Program Files\dotnet\sdk]
[22:42:25][Exec] 2.1.508 [C:\Program Files\dotnet\sdk]
[22:42:25][Exec] 3.0.100 [C:\w\99694c0b2afac045\Build\dotnetsdk\sdk]
[22:42:25][TestSolution] ............................................................
[22:42:25][TestSolution] Exec
[22:42:25][Exec] "C:\w\99694c0b2afac045\Build\dotnetsdk\dotnet" --list-runtimes
[22:42:25][Exec] Microsoft.AspNetCore.All 2.1.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
[22:42:25][Exec] Microsoft.AspNetCore.All 2.1.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
[22:42:25][Exec] Microsoft.AspNetCore.All 2.1.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
[22:42:25][Exec] Microsoft.AspNetCore.All 2.1.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
[22:42:25][Exec] Microsoft.AspNetCore.All 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
[22:42:25][Exec] Microsoft.AspNetCore.App 2.1.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
[22:42:25][Exec] Microsoft.AspNetCore.App 2.1.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
[22:42:25][Exec] Microsoft.AspNetCore.App 2.1.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
[22:42:25][Exec] Microsoft.AspNetCore.App 2.1.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
[22:42:25][Exec] Microsoft.AspNetCore.App 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
[22:42:25][Exec] Microsoft.AspNetCore.App 3.0.0 [C:\w\99694c0b2afac045\Build\dotnetsdk\shared\Microsoft.AspNetCore.App]
[22:42:25][Exec] Microsoft.NETCore.App 1.0.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
[22:42:25][Exec] Microsoft.NETCore.App 2.0.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
[22:42:25][Exec] Microsoft.NETCore.App 2.0.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
[22:42:25][Exec] Microsoft.NETCore.App 2.0.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
[22:42:25][Exec] Microsoft.NETCore.App 2.1.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
[22:42:25][Exec] Microsoft.NETCore.App 2.1.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
[22:42:25][Exec] Microsoft.NETCore.App 2.1.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
[22:42:25][Exec] Microsoft.NETCore.App 2.1.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
[22:42:25][Exec] Microsoft.NETCore.App 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
[22:42:25][Exec] Microsoft.NETCore.App 3.0.0 [C:\w\99694c0b2afac045\Build\dotnetsdk\shared\Microsoft.NETCore.App]
[22:42:25][Exec] Microsoft.WindowsDesktop.App 3.0.0 [C:\w\99694c0b2afac045\Build\dotnetsdk\shared\Microsoft.WindowsDesktop.App]

_Copied from original issue: dotnet/cli#12960_

**Comment by livarcocc** [comment]
_From @PleasantD on Tuesday, October 22, 2019 3:36:21 PM_

A little more background:
We are running compilations (via `dotnet msbuild`) and test (via `dotnet test`) on a TeamCity build server.
Our custom build script downloads and unzips a specific version of the SDK to a local build folder and runs all `dotnet` commands from that.

I had the server run `dotnet --info` and got a few more details
```
"C:\w\1eabb82de4cdcdf5\Build\dotnetsdk\dotnet" --info
.NET Core SDK (reflecting any global.json):
 Version:   3.0.100
 Commit:    04339c3a26

Runtime Environment:
 OS Name:     Windows
 OS Version:  6.3.9600
 OS Platform: Windows
 RID:         win81-x64
 Base Path:   C:\w\1eabb82de4cdcdf5\Build\dotnetsdk\sdk\3.0.100\

Host (useful for support):
  Version: 3.0.0
  Commit:  7d57652f33

.NET Core SDKs installed:
  1.0.0-preview2-003131 [C:\Program Files\dotnet\sdk]
  2.1.4 [C:\Program Files\dotnet\sdk]
  2.1.201 [C:\Program Files\dotnet\sdk]
  2.1.202 [C:\Program Files\dotnet\sdk]
  2.1.401 [C:\Program Files\dotnet\sdk]
  2.1.403 [C:\Program Files\dotnet\sdk]
  2.1.505 [C:\Program Files\dotnet\sdk]
  2.1.507 [C:\Program Files\dotnet\sdk]
  2.1.508 [C:\Program Files\dotnet\sdk]
  3.0.100 [C:\w\1eabb82de4cdcdf5\Build\dotnetsdk\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.0.0 [C:\w\1eabb82de4cdcdf5\Build\dotnetsdk\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 1.0.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.0.0 [C:\w\1eabb82de4cdcdf5\Build\dotnetsdk\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.0.0 [C:\w\1eabb82de4cdcdf5\Build\dotnetsdk\shared\Microsoft.WindowsDesktop.App]
```

**Comment by livarcocc** [comment]
_From @PleasantD on Tuesday, October 22, 2019 4:45:19 PM_

A little digging...
I am suspicious of this:
https://github.com/microsoft/vstest/blob/3a94eb15674abbee0a2cfcd13f26796f5dc6bb84/src/Microsoft.TestPlatform.Build/Tasks/VSTestForwardingApp.cs#L34

Looks like the test forwarder`VSTestForwardingApp` is starting a process to call `dotnet`.
But not the `dotnet` that was originally used, which would explain why it can't find the local runtime.

**Comment by singhsarab** [comment]
@PleasantD Could you please try setting DOTNET_ROOT environment variable to point the appropriate runtime ?

**Comment by PleasantD** [comment]
Yes, this fixes the issue.

**Comment by singhsarab** [comment]
Thanks for the confirmation.

**Comment by PleasantD** [comment]
I'm not sure why this was closed.
It still feels like a hack that we have to functionally redirect {{dotnet test}} via a PATH variable  even through we called it specifically with local version.

This should be automatic.

**Comment by ViktorHofer** [comment]
cc @vitek-karas who should know more about the host/shared framework resolution logic.

As @singhsarab mentioned, this is by design.

> Should we now consider setting the DOTNET_ROOT in the vstest runner if it is not already set ?

That could cause other unforeseeable issues. Maybe Vitek has an opinion here.

**Comment by vitek-karas** [comment]
I think there are two parts to this:
1) Which .NET Core install should the tests run on - how do we find it (where do we take this information from)
2) How to setup the environment so that the tests actually use that install

Re 1:
This depends on several things - if the tests can retarget. That is, is it allowed/enabled to run `dotnet test` using "Some" dotnet, but then the tests themselves will be using different install? Is this desirable?
I do think that if the tests are to use the same install (as in, there's nothing which tells them to run elsewhere), then I think they should prefer the install used to run the `dotnet test` itself.

Re 2:
If we're confident that in 1 we pick the right install, then setting `DOTNET_ROOT` to that location should be the right thing to do. For the tests themselves this is clearly beneficial as it gives the test runner direct control over which .NET install to run on. If the tests themselves spawn new .NET Core processes, then I think it's perfectly valid to expect those to run on the same install as well - by default.

On Windows there's also the consideration of what to do with `DOTNET_MULTILEVEL_LOOKUP`. If left unset, the host will look at `DOTNET_ROOT` first and then it will also look at the global location (Program Files). It will then pick the best match. So in the sample above it would work since 3.0 is only available in the custom install, but if 2.1 would be requested it would come from program files. The test host could also set `DOTNET_MULTILEVEL_LOOKUP=0` in which case the fallback to global locations is disabled. Not sure if that is desirable.

I personally think that if our tool (`dotnet test` in this case) spawns child .NET Core processes it should be very specific about which .NET install they run on. Relying on global install location to always get the right bits is wrong. It probably works mostly on Windows as people tend to install .NET Core to global location (but then this issues shows that it's not always the case), but it's even worse on Linux/Mac were .NET Core is actually frequently installed into custom locations.
